src/analysis/vector/geometry_checker/qgsfeaturepool.cpp

```diff
@@ -66,6 +66,9 @@ bool QgsFeaturePool::getFeature( QgsFeatureId id, QgsFeature &feature )
     }
     locker.changeMode( QgsReadWriteLocker::Write );
     mFeatureCache.insert( id, new QgsFeature( feature ) );
+
+    //cleanup the index, in case the feature was evicted from the cache but is still in the index
+    mIndex.deleteFeature( id );
```

Comment:
This doesn't do what it appears to do :scream: 

- there's no QgsSpatialIndex::deleteFeature( QgsFeatureId id ) method
- The feature ID is triggering an implicit construction of QgsFeature via the `QgsFeature( QgsFeatureID id )` constructor (which should possibly be marked explicit..., but don't touch that here!). So then QgsSpatialIndex::deleteFeature is being called with a QgsFeature which has NO geometry
- Follow the logic in QgsSpatialIndex::deleteFeature -- it requires the passed feature geometry to exist and match the geometry of the feature currently present in the index.

I think you better add some tests triggering the cache eviction logic bug first before implementing a fix :stuck_out_tongue_closed_eyes:

Reply:
I'm still -1 to introducing further known issues into this class, when we know upfront that their around and will cause someone to lose a lot of time re-discovering this and fixing in future.

Why not just go with my original proposal of adding the bounding box hash to the feature pool? That's a couple of lines of code to add, and will fix the issue properly.

Reply:
> Why not just go with my original proposal of adding the bounding box hash to the feature pool? That's a couple of lines of code to add, and will fix the issue properly.

Yes I was hoping we could use the spatial index ::geometry() method like such a hash, it seems to behave exactly like this (with FlagStoreFeatureGeometries set 😆 )

Reply:
Oh but it would store whole geometry data rather than only small bounding boxes... 

Ok let's go with the hash, I have the impression to triple the data with feature cache, index, bounding boxes, but I agree that it is necessary in the end (and I have learned a lot 👍🏼  thanks)